### PR TITLE
add support for podmonitor

### PR DESCRIPTION
**Pull request by wei-lee**
## JIRA ID

https://issues.redhat.com/browse/INTLY-4129

## Additional Information

This PR adds another monitor resource: PodMonitor. It is useful when there are multiple Keycloak pods are running at the same time in a cluster (HA mode).

## Checklist:
- [ ] Verified by team member
- [ ] Comments where necessary
- [ ] Automated Tests
- [ ] Documentation changes if necessary

### CODE REVIEW

File: deploy/crds/keycloak.org_keycloaks_crd.yaml [hunk 1]
```diff
@@ -69,6 +69,7 @@ spec:
                   type: boolean
               type: object
             instances:
+              default: 1
```
Comment (slaskawi) [inline deploy/crds/keycloak.org_keycloaks_crd.yaml]: It seems this approach doesn't work correctly. Otherwise, Travis build should probably report success: https://travis-ci.org/keycloak/keycloak-operator/builds/622755328?utm_source=github_status&utm_medium=notification

File: go.mod [hunk 2]
```diff
@@ -2,61 +2,69 @@ module github.com/keycloak/keycloak-operator
 
 require (
 	cloud.google.com/go v0.46.3 // indirect
-	github.com/blang/semver v3.5.1+incompatible // indirect
+	github.com/bombsimon/wsl v1.2.8 // indirect
 	github.com/coreos/prometheus-operator v0.29.0
-	github.com/go-openapi/spec v0.19.0
-	github.com/gogo/protobuf v1.3.0 // indirect
+	github.com/go-critic/go-critic v0.4.0 // indirect
+	github.com/go-openapi/spec v0.19.2
```
Comment (slaskawi) [inline go.mod]: It seems you ran `go mod tidy` or `go mod vendor`. Could you please put vendoring changes into a separate commit? This way, reviewing will be much easier.

File: pkg/apis/keycloak/v1alpha1/zz_generated.openapi.go [hunk 3]
```diff
@@ -11,22 +11,22 @@ import (
...
-		"./pkg/apis/keycloak/v1alpha1.KeycloakRealm":        schema_pkg_apis_keycloak_v1alpha1_KeycloakRealm(ref),
-		"./pkg/apis/keycloak/v1alpha1.KeycloakRealmSpec":    schema_pkg_apis_keycloak_v1alpha1_KeycloakRealmSpec(ref),
-		"./pkg/apis/keycloak/v1alpha1.KeycloakRealmStatus":  schema_pkg_apis_keycloak_v1alpha1_KeycloakRealmStatus(ref),
-		"./pkg/apis/keycloak/v1alpha1.KeycloakSpec":         schema_pkg_apis_keycloak_v1alpha1_KeycloakSpec(ref),
-		"./pkg/apis/keycloak/v1alpha1.KeycloakStatus":       schema_pkg_apis_keycloak_v1alpha1_KeycloakStatus(ref),
-		"./pkg/apis/keycloak/v1alpha1.KeycloakUser":         schema_pkg_apis_keycloak_v1alpha1_KeycloakUser(ref),
-		"./pkg/apis/keycloak/v1alpha1.KeycloakUserSpec":     schema_pkg_apis_keycloak_v1alpha1_KeycloakUserSpec(ref),
-		"./pkg/apis/keycloak/v1alpha1.KeycloakUserStatus":   schema_pkg_apis_keycloak_v1alpha1_KeycloakUserStatus(ref),
+		"github.com/keycloak/keycloak-operator-1/pkg/apis/keycloak/v1alpha1.Keycloak":             schema_pkg_apis_keycloak_v1alpha1_Keycloak(ref),
+		"github.com/keycloak/keycloak-operator-1/pkg/apis/keycloak/v1alpha1.KeycloakAWSSpec":      schema_pkg_apis_keycloak_v1alpha1_KeycloakAWSSpec(ref),
```
Comment (slaskawi) [inline pkg/apis/keycloak/v1alpha1/zz_generated.openapi.go]: This doesn't look right. Could you please revert this changes?

File: pkg/controller/keycloak/keycloak_controller.go [hunk 4]
```diff
@@ -108,11 +108,16 @@ func add(mgr manager.Manager, r reconcile.Reconciler, autodetectChannel chan sch
 				common.WatchSecondaryResource(c, ControllerName, gvk.Kind, &monitoringv1.PrometheusRule{}, &kc.Keycloak{}) // nolint
 			}
 
-			// Check if this channel event was for the ServiceMonitor resource type
+			// Check if this channel event was for the PodMonitor resource type
```
Comment (slaskawi) [inline pkg/controller/keycloak/keycloak_controller.go]: Probably a mass replace mistake.

File: pkg/controller/keycloak/keycloak_reconciler.go [hunk 5]
```diff
@@ -196,6 +197,30 @@ func (i *KeycloakReconciler) GetKeycloakServiceMonitorDesiredState(clusterState
...
+		return common.GenericCreateAction{
+			Ref: podmonitor,
+			Msg: "create keycloak service monitor",
+		}
+	}
+
+	podmonitor.ResourceVersion = clusterState.KeycloakPodMonitor.ResourceVersion
+	return common.GenericUpdateAction{
+		Ref: podmonitor,
+		Msg: "update keycloak service monitor",
```
Comment (slaskawi) [inline pkg/controller/keycloak/keycloak_reconciler.go]: Probably should be "update keycloak pod monitor"

File: pkg/controller/keycloak/keycloak_reconciler.go [hunk 6]
```diff
@@ -196,6 +197,30 @@ func (i *KeycloakReconciler) GetKeycloakServiceMonitorDesiredState(clusterState
...
+	if !keyExists || !resourceWatchExists {
+		return nil
+	}
+
+	podmonitor := model.PodMonitor(cr)
+
+	if clusterState.KeycloakPodMonitor == nil {
+		return common.GenericCreateAction{
+			Ref: podmonitor,
+			Msg: "create keycloak service monitor",
```
Comment (slaskawi) [inline pkg/controller/keycloak/keycloak_reconciler.go]: Probably should be "update keycloak pod monitor"

File: go.mod [hunk 7]
```diff
@@ -2,61 +2,69 @@ module github.com/keycloak/keycloak-operator
...
 	github.com/grpc-ecosystem/grpc-gateway v1.11.2 // indirect
 	github.com/hashicorp/golang-lru v0.5.3 // indirect
 	github.com/integr8ly/grafana-operator v1.4.0
 	github.com/json-iterator/go v1.1.7
-	github.com/onsi/ginkgo v1.10.1 // indirect
-	github.com/onsi/gomega v1.7.0 // indirect
+	github.com/mattn/go-isatty v0.0.10 // indirect
 	github.com/openshift/api v3.9.0+incompatible
-	github.com/operator-framework/operator-sdk v0.10.1-0.20190911145116-334c667503d0
+	github.com/operator-framework/operator-sdk v0.0.0-00010101000000-000000000000
```
Comment (slaskawi) [inline go.mod]: This one looks very, very suspicious!

File: go.mod [hunk 8]
```diff
@@ -2,61 +2,69 @@ module github.com/keycloak/keycloak-operator
...
 	google.golang.org/grpc v1.24.0 // indirect
-	gopkg.in/check.v1 v1.0.0-20190902080502-41f04d3bba15 // indirect
-	k8s.io/api v0.0.0-20190612125737-db0771252981
-	k8s.io/apimachinery v0.0.0-20190612125636-6a5db36e93ad
-	k8s.io/client-go v11.0.0+incompatible
-	k8s.io/kube-openapi v0.0.0-20190603182131-db7b694dc208
+	gopkg.in/yaml.v2 v2.2.7 // indirect
+	k8s.io/api v0.0.0-20190918155943-95b840bb6a1f
+	k8s.io/apimachinery v0.0.0-20190913080033-27d36303b655
+	k8s.io/client-go v12.0.0+incompatible
```
Comment (slaskawi) [inline go.mod]: It seems you upgraded Kubernetes version here as well? What about @davidffrench https://github.com/keycloak/keycloak-operator/pull/111 ?

**Comment by slaskawi** [comment]
Thanks for the Pull Request @wei-lee ! There's a Travis failure to be fixed and it seems you accidentally messed up dependencies. Perhaps you should put your changes on top of https://github.com/keycloak/keycloak-operator/pull/111?

**Comment by davidkirwan** [comment]
Folks I don't have the ability to push changes to Wei's PR here. I've created another PR from @wei-lee 's PR (https://github.com/keycloak/keycloak-operator/pull/114) and will attempt to deal with the suggestions made in this PR.

**Comment by slaskawi** [comment]
@stianst @abstractj May I ask you guys to close this PR?
